src/plugins/newsfeed/public/lib/driver.ts

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import moment from 'moment';
+import { HttpSetup } from 'kibana/public';
+import * as Rx from 'rxjs';
+import { NEWSFEED_DEFAULT_SERVICE_BASE_URL } from '../../common/constants';
+import { ApiItem, FetchResult, NewsfeedPluginBrowserConfig } from '../types';
+import { convertItems } from './convert_items';
+import type { NewsfeedStorage } from './storage';
+
+type ApiConfig = NewsfeedPluginBrowserConfig['service'];
+
+export class NewsfeedApiDriver {
+  private readonly kibanaVersion: string;
+  private readonly loadedTime = moment().utc(); // the date is compared to time in UTC format coming from the service
+
+  constructor(
+    kibanaVersion: string,
+    private readonly userLanguage: string,
+    private readonly fetchInterval: number,
+    private readonly storage: NewsfeedStorage
+  ) {
+    // The API only accepts versions in the format `X.Y.Z`, so we need to drop the `-SNAPSHOT` or any other label after it
+    this.kibanaVersion = kibanaVersion.replace(/^(\d+\.\d+\.\d+).*/, '$1');
+  }
+
+  shouldFetch(): boolean {
+    const lastFetchUtc = this.storage.getLastFetchTime();
+    if (!lastFetchUtc) {
+      return true;
+    }
+    const last = moment(lastFetchUtc);
+
+    // does the last fetch time precede the time that the page was loaded?
+    if (this.loadedTime.diff(last) > 0) {
+      return true;
+    }
+
+    const now = moment.utc(); // always use UTC to compare timestamps that came from the service
+    const duration = moment.duration(now.diff(last));
+    return duration.asMilliseconds() > this.fetchInterval;
+  }
+
+  fetchNewsfeedItems(http: HttpSetup, config: ApiConfig): Rx.Observable<FetchResult> {
+    const urlPath = config.pathTemplate.replace('{VERSION}', this.kibanaVersion);
+    const fullUrl = (config.urlRoot || NEWSFEED_DEFAULT_SERVICE_BASE_URL) + urlPath;
+
+    return Rx.from(
+      http
+        .fetch(fullUrl, {
+          method: 'GET',
+        })
```

Comment:
Should we really be using the HttpService here? It's only intended for query Kibana's backend but this queries an external URL. It could lead to bugs such as the user being logged out if this service returns a 403 for any reason.

I filed an issue to prevent this type of usage here: https://github.com/elastic/kibana/issues/100803

It also appears there's no unit tests on this function.

Reply:
> Should we really be using the HttpService here

Seeing https://github.com/elastic/kibana/issues/100803, I guess probably not. However, what alternative should we recommend here to query remote endpoints from the clientside?

However, given that this call (and the implementation) was present before current PR, do we want to do that now, or should that be done as a follow-up once we're settled on the alternative we want to officially recommend?

Reply:
Migrated to fetch in https://github.com/elastic/kibana/pull/100357/commits/3679e941ebb865409182fa78093fbe5f76b0b648